File: 프로그래머스/CodingTest/DynamicProgramming/004.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

bool cmp(int a, int b){
    return a > b;
}

int solution(vector<int> money) {
    int answer = 0;
    vector<int> v;
    vector<bool> w;
    if(money.size() == 3){
        sort(money.begin(), money.end(), cmp);
        answer = money[0];
        return answer;
    }
    for(int i=0;i<2;i++){
        v.push_back(money[i]);
    }
    w.push_back(1);
    w.push_back(0);
    w.push_back(1);
    v.push_back(money[2] + money[0]);
    for(int i=3;i<money.size();i++){
        int max = v[i-3] + money[i];
        bool f = w[i-3];
        if(v[i-3] < v[i-2]){
            max = money[i] + v[i-2];
            f = w[i-2];
        }
        v.push_back(max);
    }
    if(w.back() == 1){
        answer = v[v.size()-2];
    }else{
        if(v[v.size()-1] > v[v.size()-1]){
            answer = v[v.size()-1];
        }else{
            answer = v[v.size()-2];
        }
    }
    return answer;
}

// 11 0 13 16 113 116 198 117

File: 프로그래머스/weeklyChallenge/004.js
function solution(table, languages, preference) {
    var answer = '';
    var tableTmp = [];
    var scores = [];
    for(var i=0;i<table.length;i++){
        var tmp = table[i].split(' ');
        tableTmp.push(tmp);
    }
    //console.log(tableTmp);
    for(var i=0;i<tableTmp.length;i++){
        var scoreTmp = 0;
        for(var j=1;j<=5;j++){
            for(var k=0;k<languages.length;k++){
                var check = false;
                if(tableTmp[i][6-j] == languages[k]){
                    scoreTmp = scoreTmp + preference[k] * j;
                    check = true;
                    break;
                }
            }
            
            //console.log(tableTmp[i][6-j]);
        }
        scores.push(scoreTmp);
        //console.log('');
    }
    //console.log(scores);
    var max = 0;
    var maxScore = [];
    for(var i=0;i<scores.length;i++){
        if(scores[i] > max){
            max = scores[i];
            maxScore = [];
            maxScore.push(tableTmp[i][0]);
        }else if(scores[i] == max){
            maxScore.push(tableTmp[i][0]);
        }
    }
    if(maxScore.length == 1){
        answer = maxScore[0];
    }else{
        maxScore.sort();
        answer = maxScore[0];
    }
    //console.log(maxScore);
    return answer;
}